tools/brokerd/service/win32/broker_service.cpp

```diff
@@ -36,35 +181,93 @@ sample_service::~sample_service () {
 /// \note start_handler() must return to the operating system after the service's operation has
 /// begun: it must not block.
 
-void sample_service::start_handler (DWORD /*argc*/, TCHAR * /*argv*/[]) {
+void broker_service::start_handler (DWORD argc, TCHAR * argv[]) {
     // Log a service start message to the Application log.
-    this->write_event_log_entry ("CppWindowsService in OnStart", event_type::information);
-    thread_ = std::thread ([this] () { this->worker (); });
+    this->write_event_log_entry ("broker service starting", event_type::information);
+
+    std::pair<switches, int> opts = get_switches (argc, argv);
```

Comment:
This call to `get_switches()` results in an ACCESS_VIOLATION which I think is due to the fact that `main.cpp` uses the same argument parsing functionality.

Reply:
If you can explain how I can reproduce the crash, it shouldn’t be hard to figure out.

For background: since we support both “inside LLVM” and “standalone” builds for pstore, I needed to have command-line parsing for both. My original idea was to use the LLVM command-line parser when available and to have an API-compatible library for the standalone. 

This didn’t work well (‘cause Windows UNICODE mode), so I now use the API-compatible version all the time. Of course, this means that I no longer need to be API-compatible.

The LLVM-style command-line parser has a lot of global state (I’m not a fan of that and getting rid of it is on my list, but will touch a lot of code). Calling the parser more than once from a process can be made to work: the unit tests do exactly that.

Reply:
The stack trace is as follows:

```C++
>	pstore-broker-service.exe!pstore::command_line::details::parse_positional_arguments::__l2::<lambda>(const pstore::command_line::option * const opt) Line 293	C++
 	[External Code]	
 	pstore-broker-service.exe!pstore::command_line::details::parse_positional_arguments<...>(std::_Vector_iterator<...> first_arg, std::_Vector_iterator<...> last_arg) Line 297	C++
 	pstore-broker-service.exe!pstore::command_line::details::parse_command_line_options<...>(std::_Vector_iterator<...> first_arg, std::_Vector_iterator<...> last_arg, const std::string & overview, std::basic_ostream<...> & outs, std::basic_ostream<...> & errs) Line 325	C++
 	pstore-broker-service.exe!pstore::command_line::parse_command_line_options<...>(const int argc, wchar_t * const * argv, const std::string & overview) Line 345	C++
 	pstore-broker-service.exe!get_switches(int argc, wchar_t * * argv) Line 67	C++
 	pstore-broker-service.exe!broker_service::start_handler(unsigned long argc, wchar_t * * argv) Line 74	C++
 	pstore-broker-service.exe!wmain(int argc, wchar_t * * argv) Line 67	C++
 	[External Code]	

```
You can trigger the crash by either installing the service and running it (extremely hard to debug though) or you can apply the following patch and run the service as a normal console app:
```diff
diff --git a/tools/brokerd/service/win32/broker_service.hpp b/tools/brokerd/service/win32/broker_service.hpp
index d2b34859..b7c073d2 100644
--- a/tools/brokerd/service/win32/broker_service.hpp
+++ b/tools/brokerd/service/win32/broker_service.hpp
@@ -37,7 +37,7 @@ public:
diff --git a/tools/brokerd/service/win32/broker_service.hpp b/tools/brokerd/service/win32/broker_service.hpp
index d2b34859..b7c073d2 100644
--- a/tools/brokerd/service/win32/broker_service.hpp
+++ b/tools/brokerd/service/win32/broker_service.hpp
@@ -37,7 +37,7 @@ public:
                     bool can_pause_continue = false);
     ~broker_service () override;

-protected:
+public:
     void start_handler (DWORD argc, TCHAR * argv[]) override;
     void stop_handler () override;

diff --git a/tools/brokerd/service/win32/main.cpp b/tools/brokerd/service/win32/main.cpp
index d50339fb..7ce3e734 100755
--- a/tools/brokerd/service/win32/main.cpp
+++ b/tools/brokerd/service/win32/main.cpp
@@ -41,8 +41,8 @@ namespace {

     using namespace pstore::command_line;

-    opt<bool> install_opt ("install", desc ("Install the service"));
-    opt<bool> remove_opt ("remove", desc ("Remove the service"));
+    //opt<bool> install_opt ("install", desc ("Install the service"));
+    //opt<bool> remove_opt ("remove", desc ("Remove the service"));

 } // end anonymous namespace

@@ -52,20 +52,20 @@ int _tmain (int argc, TCHAR * argv[]) {
     try {
         parse_command_line_options (argc, argv, "pstore broker server");

-        if (install_opt && remove_opt) {
-            std::wcerr << L"--install and --remove cannot be specified together!\n";
-            std::exit (EXIT_FAILURE);
-        }
+        //if (install_opt && remove_opt) {
+        //    std::wcerr << L"--install and --remove cannot be specified together!\n";
+        //    std::exit (EXIT_FAILURE);
+        //}

-        if (install_opt) {
-            install_service (service_name, display_name, start_type, service_dependencies, account,
-                             account_password);
-        } else if (remove_opt) {
-            uninstall_service (service_name);
-        } else {
+        //if (install_opt) {
+        //    install_service (service_name, display_name, start_type, service_dependencies, account,
+        //                     account_password);
+        //} else if (remove_opt) {
+        //    uninstall_service (service_name);
+        //} else {
             broker_service service{service_name};
-            broker_service::run (service);
-        }
+            service.start_handler (argc, argv);
+        //}
     } catch (std::exception const & ex) {
         std::wcerr << L"Error: " << pstore::utf::to_native_string (ex.what ()) << std::endl;
         exit_code = EXIT_FAILURE;

```